Update guide.md

ndilekli opened the pull request:
Naci's SWC homework

---

Thanks for contributing! If this contribution is for instructor training, please send an email to checkout@carpentries.org with a link to this contribution so we can record your progress. You’ve completed your contribution step for instructor checkout just by submitting this contribution.  

Please keep in mind that lesson maintainers are volunteers and it may be some time before they can respond to your contribution. Although not all contributions can be incorporated into the lesson materials, we appreciate your time and effort to improve the curriculum.  If you have any questions about the lesson maintenance process or would like to volunteer your time as a contribution reviewer, please contact The Carpentries Team at team@carpentries.org.  

---

ldko commented:
Hello @ndilekli , thank you for submitting an update for the guide. I see you have opened two conflicting pull requests modifying the language of guide.md. Is this pull request meant to supercede #836 ? If you could close one of the PR's, it will inform which one we should review. Also, could you and a short explanation to the PR to explain what your goals for updating guide.md are. This will also help with the review. Thanks again!

ndilekli commented:
Goals: 1) Further clarify why Python is a preferred programming language, 2) The existing guide sounded like functions are the most important part of a programming language, I changed the language a little and also provided an explanation why functions are very important (reusability and organization); and 3) Added comments about a common user error when dealing with multiple cells in Jupyter.

ndilekli commented:
Hello yes, I am quite a novice with teamwork with git. I removed #836, and
added an explanation of the goals in the comments section. Please let me
know if that will not work and where else I can put that information.
Thanks!

On Mon, Jun 22, 2020 at 6:14 PM Lauren Ko <notifications@github.com> wrote:

> Hello @ndilekli <https://github.com/ndilekli> , thank you for submitting
> an update for the guide. I see you have opened two conflicting pull
> requests modifying the language of guide.md. Is this pull request meant to
> supercede #836
> <https://github.com/swcarpentry/python-novice-inflammation/pull/836> ? If
> you could close one of the PR's, it will inform which one we should review.
> Also, could you and a short explanation to the PR to explain what your
> goals for updating guide.md are. This will also help with the review.
> Thanks again!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/swcarpentry/python-novice-inflammation/pull/837#issuecomment-647582035>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA66AAOGH73U2MGC6KDOOTDRX5YOFANCNFSM4OEQOHBQ>
> .
>

ldko commented on _extras/guide.md:
```diff
@@ -26,9 +26,13 @@ is included as [`code/gen_inflammation.py`](../code/gen_inflammation.py).
+variables, conditionals, looping and functions.
+We make a special effort to introduce the functions, the single most important idea in programming
+that allows reusability and efficient logical organization of the software.
+We show how to solve problems by building functions,
```
I wasn't fond of the original language because it is so strongly worded with saying "the single most important idea in programming". I like that you add in a reason for why functions are important. However, I think the new wording takes some of the focus off of problem solving and puts it more on concepts of programming, where I think we want to highlight that we are teaching learners to solve problems with code, not about the code itself. I suggest tweaking the language to put the focus back on problem solving while retaining your explanation of why functions are important. What do you (and anyone else who cares to share) think about:
```
This lesson is written as an introduction to Python,
but its real purpose is to introduce an important idea in programming:
how to solve problems by building functions,
units of code that allow reusability and efficient logical organization of software.
In order to teach that,
we must teach people a little about
programming concepts such as variables, conditionals, loops,
and the mechanics of manipulating data with lists and file I/O
so that their functions can do things they actually care about.
```

slel replied:
I like your suggestions. Maybe "file I/O" can be a little intimidating?
How about the following:

```
This lesson is written as an introduction to Python,
but its wider purpose is to introduce an important idea in programming:
how to solve problems by building functions, units of code that allow
reusability and efficient logical organization of software.

So that our functions can do things we actually care about,
we cover programming concepts such as variables, conditionals,
loops, and the mechanics of manipulating data with lists,
and of reading from files and writing to files.
```

ndilekli replied:
I would agree with both points, to make a special emphasis on functions, and to perhaps not to intimidate the potentially non-technical crowds with terms like the file I/O operations. I like @slel's version, since it incorporates both improvements. I should note though, that the training did not have the feeling like there was a particular effort on the introduction and practice of functions. If that's the intent, perhaps the curriculum might need updating sometime as well.

ldko replied:
I agree with @slel (thanks for commenting!) about replacing the term "file I/O", though I don't think we really spend time on writing to files. Maybe the second sentence could be:
```
So that our functions can do things we actually care about,
we cover programming concepts such as variables, conditionals,
loops, lists, and the mechanics of manipulating data read from files.
```

ndilekli replied:
Very good, for the sake of simplicity I suggest: 
```suggestion
So that our functions can do things we actually care about,
we cover programming concepts such as variables, conditionals,
loops, lists, and reading data from files.
```

ldko commented on _extras/guide.md:
```diff
@@ -52,7 +56,8 @@ Explain that we use Python because:
 *   It has a large scientific user community.
-*   It's easier for novices to learn than most of the mature alternatives.
+*   It's easier for novices to learn than most of the mature alternatives, 
+especially due to its syntax that reads like a human language.
```
What do you think of toning this down a little to say:
```
*   It's easier for novices to learn than most of the mature alternatives, 
partially due to its syntax reading more like a human language.
```

slel replied:
partly?

ldko commented on _extras/guide.md:
```diff
@@ -67,6 +72,9 @@ Instructors should, however, walk learners through some basic operations:
 *   Explain the difference between `In[#]` and `Out[#]`.
+*   Explain that the cells are independent blocks 
+and if we make a change at a top cell we need to run all the subsequent
+cells to make sure the change is reflected across the script.
```
I think this is a concept worth including, as it often comes up in workshops. I do wonder, though, if teaching it at the very beginning of a lesson might start to overload learners and be confusing to see before people new to programming have seen much code. If people want to include it, I suggest possibly rewriting it more as an operation:
```
* Show that modifying code in an earlier cell requires re-running any subsequent
cell in order for it to reflect the change.
```

ndilekli replied:
Much better wording, thanks! I think you point out to an interesting concern: perhaps this guide can be eventually restructured so that it roughly explains which points correspond to which portions of the curriculum because some particular issues like this one can be confusing at the beginning, but they are the things that better be provided to the students at some point in a curriculum with such scope with fairly complex workflows. To keep things relatively simple, we could flag some sentences in the following way:

```suggestion
* After few episodes show that modifying code in an earlier cell requires re-running any subsequent cells in order for it to reflect the change.
```

ldko replied:
I don't think it needs to be a few episodes in, maybe leaving it open-ended is ok, allowing the instructor to teach it when there are adequate cells for the concept. We should be sure to match plural (cells/them) though:
```
* Show that modifying code in an earlier cell requires re-running any
subsequent cells in order for them to reflect the change.
```

ndilekli commented:
A question: Am I the one that will need to mark "Resolve conversation" once things look well, or will that be you @ldko ?

ldko commented:
Good question @ndilekli :). In most cases it tends to be the maintainers that mark the conversations resolved. I tend to prefer to do that after commits are pushed to the branch/PR that address the conversation when that is relevant. I think the way things are now is fine, but let's see if anyone else from the community has any suggestions (let's give people a day or so), and then if you would go ahead and commit the decided upon changes, that would be great. Thank you very much for participating in the process!

maxim-belkin commented:
> let's give people a day or so

No need to wait, I'm here :)

Hi, Naci! Welcome to the Carpentries! I'm a co-maintainer on this lesson. Thank you for bringing this up and working on this. 

I like your changes and all suggestions (Lauren's and Samuel's). And it looks like you mostly agree with them. So, because you're new to working with Git, I'd like to clarify that we (maintainers) review proposed changes and propose modifications and contributors (in this case this is you) are expected to update their pull requests if they agree with the proposed modifications. So, if you agree with the proposed changes -- please go ahead and make these modifications. You can do all that from your browser window --  navigate to https://github.com/ndilekli/python-novice-inflammation/blob/patch-2/_extras/guide.md, click the Pencil (Edit this file) icon, make the changes, then at the very bottom of that page describe your changes ("Update based on maintainers' feedback") and commit directly to patch-2 branch.

As it currently stands, I approve of the changes proposed in this PR provided that requested changes are implemented.

Again, thank you for contributing to the lesson! This is looking great so far! 👍

ndilekli commented:
> navigate to https://github.com/ndilekli/python-novice-inflammation/blob/patch-2/_extras/guide.md, click the Pencil (Edit this file) icon, make the changes, then at the very bottom of that page describe your changes ("Update based on maintainers' feedback") and commit directly to patch-2 branch.

Hmm... I tried that but the 'Commit Changes" button is disabled.

I also notice that some of the improvements are not yet incorporated to this branch such as:

> * Show that modifying code in an earlier cell requires re-running any
subsequent cells in order for them to reflect the change.

maxim-belkin commented:
> Hmm... I tried that but the 'Commit Changes" button is disabled.

You need to type a title (such as "Update based on received feedback" or something more descriptive -- your choice) -- it should become enabled as soon as you do that.

> I also notice that some of the improvements are not yet incorporated to this branch such as:
> 
> > * Show that modifying code in an earlier cell requires re-running any
> >   subsequent cells in order for them to reflect the change.

That was Lauren's suggestion. You need to add it to your branch as I described above. Unless what you're saying is that you did and it doesn't show up. If you did it on your computer, make sure:
- you added and committed your changes to `patch-2` branch
- pushed that branch to github with `git push https://github.com/ndilekli/python-novice-inflammation.git patch-2`

ndilekli commented:
> You need to type a title (such as "Update based on received feedback" or something more descriptive -- your choice) -- it should become enabled as soon as you do that.

I am still having the same problem. I tried with Chrome and Opera.
![commit changes disabled](https://user-images.githubusercontent.com/4055041/85924379-9d321180-b89a-11ea-832d-d33335af6df4.png)
 with Chrome and Opera.

maxim-belkin commented:
Looks like you were able to make one update -- https://github.com/swcarpentry/python-novice-inflammation/pull/837/commits/26b9528fdeba7984f581a87bfb713c9e8d690cf4

Do you remember what you did when you made that change?

ndilekli commented:
> Looks like you were able to make one update -- [26b9528](https://github.com/swcarpentry/python-novice-inflammation/commit/26b9528fdeba7984f581a87bfb713c9e8d690cf4)
> 
> Do you remember what you did when you made that change?

I think all I did was to resolve, thinking that would make the suggested changes (by Lauren and Samuel) get incorporated into the version.

ldko commented:
@ndilekli , when I go to your branch at  https://github.com/ndilekli/python-novice-inflammation/blob/patch-2/_extras/guide.md and click on the pencil icon, at first I see at the bottom that the `Commit changes` button is disabled. However, I noticed when I click in the file editor (above the "Commit changes" form) and start modifying the text of guide.md, the `Commit changes` button becomes enabled (even if I haven't entered a title yet). Can you confirm if you have made edits to the guide.md text before trying to commit the changes?

ndilekli commented:
@ldko: Very good, the commit worked [768817a](https://github.com/ndilekli/python-novice-inflammation/commit/768817a742eb80db2e5f3ccfeae112a29cc91b3d)! I needed to make some manual changes in the file so that it would count as a commit, which makes sense in retrospect. I also noticed that there were some broken lines (due to my commenting most probably), which I tried to fix in this manual version as well.

So I guess whenever a conversation is resolved with some suggestions, that in itself becomes a commit, is that correct?

Let me know if you would like me to do anything else.

Thanks, Naci

ldko commented on _extras/guide.md:
```diff
@@ -52,7 +53,8 @@ Explain that we use Python because:
 *   It has a large scientific user community.
-*   It's easier for novices to learn than most of the mature alternatives.
+*   It's easier for novices to learn than most of the mature alternatives, 
+especially due to its syntax that reads like a human language.
```
```suggestion
partly due to its syntax that reads more like a human language.
```

ldko replied:
Hi Naci, that's great you got the changes committed to the branch for the pull request! Regarding this question:

> So I guess whenever a conversation is resolved with some suggestions, that in itself becomes a commit, is that correct?

Clicking the `Resolve conversation` button on a discussion point does not commit anything to your files; it is a mechanism of the GitHub interface to say, "ok, we've made a decision on this and no longer need to discuss it further."

When the maintainers do a review of your pull request, we tend to make comments that suggest ideas for editing your submission, and GitHub allows us to put the comment in-line with the code, so we can all see what we are referencing. Generally this does not automatically update the code, it is giving you our opinion about the code change, and we want you to think about if you want to do this then manually change the file either in the GitHub interface (as you did), or at the command-line using `git`.

It is also possible to make a "suggestion" when doing review of the pull request, where we are actually asking if you want to commit the exact change we are proposing. That is what I have done above with the "partly due to its syntax..." comment, and you should be able to click on the `Commit suggestion` dropdown to commit the changes if you agree with them. In this case of using the `Commit suggestion` button, you do not need to manually edit a file, GitHub will automatically make the commit for you. I believe you used this feature earlier on this pull request. If we did not click on `Commit suggestion` the change would not get automatically committed. If you would like to included the "Suggested change" I have added above this, please try committing the suggestion with the interface above to see how that works. :)

Does that clarify these different aspects of the pull request process?

ndilekli replied:
Lauren, thank you very much, I think I have a much better understanding of the process now. I think I have a relatively steep learning curve since I never worked on SW development in teams, but I can say it sure is a very powerful process based on all these elaborate schemes to put all those inputs and interactions into the development.

ldko commented on _extras/guide.md:
```diff
@@ -67,6 +69,7 @@ Instructors should, however, walk learners through some basic operations:
 *   Create a new notebook.
 *   Enter code or data in a cell and execute it.
 *   Explain the difference between `In[#]` and `Out[#]`.
+*   Show that modifying code in an earlier cell requires re-running any subsequent cells in order for it to reflect the change.
```
One more thing @ndilekli , we try to keep the lines less than 100 characters in the markdown (see #498 for explanation). Here I am suggesting the change to shorten the line and to change "it" to "them". Thanks again for all your work on this! It's good to know it is teaching you something about collaborating on GitHub.
```suggestion
*   Show that modifying code in an earlier cell requires re-running any subsequent
    cells in order for them to reflect the change.
```

ndilekli replied:
Sure thing!

maxim-belkin commented on _extras/guide.md:
```diff
@@ -27,13 +27,14 @@ is included as [`code/gen_inflammation.py`](../code/gen_inflammation.py).
+how to solve problems by building functions, units of code that allow
+reusability and efficient logical organization of software.
+
+So that our functions can do things we actually care about,
+we cover programming concepts such as variables, conditionals,
```
"So that our functions can do things we actually care about,"  would've been a great intro to a paragraph describing how to write better functions when we already know how to write simple ones. I think we're one or two steps away from there. Would any of the following fit better the prose of this section?:

> To show learners how to write functions that do things they care about, we cover basic programming concepts such as variables, ...

or
> To teach how write functions that do exactly what we care about, we cover basic programming concepts such as variables, ...

ldko replied:
In the context of the Instructor Notes and this being just after a statement about what functions are for, I don't really think the prose options above will make for much more of a better fit for the section. However, I prefer your first option to the second (because I don't favor the use of "exactly") if others feel the wording should be changed.

maxim-belkin replied:
> (because I don't favor the use of "exactly")

Yeah, I think I agree with that.

Anyways, my point was that we cover programming concepts that we cover not to make our functions do things we care about but to teach/learn how to program in Python. We also learn about functions (which should be added to the list of things we learn). So, here is another possible justification:

> To teach how to write Python programs that perform calculations or carry out analysis that we need, we cover most important programming concepts such as variables, conditional statements, loops, lists, functions. We also show how to read data from files and provide input on the command line.

ldko replied:
Perhaps you are speaking to the point @ndilekli brought up earlier:

> the training did not have the feeling like there was a particular effort on the introduction and practice of functions.

In this case, perhaps we should take the emphasis off of functions in the previous sentence as well and say something along the lines of:

> This lesson is written as an introduction to Python,
but, its broader purpose is to introduce how to solve problems by developing programs.
By writing lines of instructions to carry out tasks, we automate our work and
enable reproducible results.
To teach how to write Python programs that perform calculations or carry out analysis that we need,
we cover most important programming concepts such as variables, conditional statements, loops,
lists, and functions.
We also show how to read data from files and provide input on the command line.

maxim-belkin replied:
Yes, I like this paragraph a lot more: it's easy to read and makes the right emphasis (I'm not sure about "lines of instructions" in the second sentence -- perhaps we need to clearly state that programs are lines of code/instructions -- but that's a minor detail).

maxim-belkin commented on _extras/guide.md:
```diff
@@ -52,7 +53,8 @@ Explain that we use Python because:
 *   It has a lot of scientific libraries, and more are constantly being added.
 *   It has a large scientific user community.
-*   It's easier for novices to learn than most of the mature alternatives.
+*   It's easier for novices to learn than most of the mature alternatives, 
+partly due to its syntax that reads more like a human language.
```
I agree that we need to explain why [we think] Python is easier to learn [than other programming languages]. However, these bullet points state the reasons why [we think] Python is better suited for teaching basic Programming skills. So, I think it might be better to not explain ourselves right here but move the explanation to the next paragraph. It could start with something like:

> That last point above is important. Code written in Python reads more like a human language. Originally, Software Carpentry was teaching programming using Perl. When we switched to Python, we found..."

What do you guys think?

ldko replied:
I think it is fine as-is. I think it will be easier to miss the point if it is moved to a separate paragraph. Also, the lead-in to the list starts with "Explain", so I think it is okay to explain in the list.

maxim-belkin replied:
> I think it will be easier to miss the point if it is moved to a separate paragraph.

I'm not suggesting to move the entire point to a separate paragraph -- only its justification / explanation. Keeping only the main points in the list should make them easier to remember. Those who don't read the next paragraph will miss the explanation(s) but not the point(s). 

Speaking of learning... I believe it's the combination of syntax and semantics that makes Python easy to learn: syntax is related to writing proper Python code, and semantics -- to making sense of Python code.

> Also, the lead-in to the list starts with "Explain"

To me this serves as a justification to write bullet points as succinct statements rather than sentences. Explanations that we can provide in the list can only be of limited length and I think we shouldn't limit our explanations by the choice of where we provide them.

Do you think the following would convey the message of the last bullet point?

> Its easy-to-learn syntax and semantics make it ideal for teaching basic programming skills.

ldko replied:
Taking out from the bullet point the part about Perl as a  justification that Python is easier, now we would be stating Python is "easy-to-learn" which I don't like. What makes for Python being "easy-to-learn syntax and semantics"? It may not be considered easy to someone, especially someone who has no reference to other programming syntax. How are the semantics of Python different than other languages?

maxim-belkin replied:
> Taking out from the bullet point the part about Perl as a justification that Python is easier, now we would be stating Python is "easy-to-learn" which I don't like. 

I guess my last suggestion was not the best and it confused you. Let me try to better explain my concern. 
What we have now in this PR can be represented as:

> we do X because:
> A
> B
> C, partially due to D (E)

This last bullet point seems to be  overloaded with information a little bit, so I propose to say:

> we do X because
> A
> B
> C
>
> C is partially due to D which, in turn, is supported by E.

(A, B, C) together explain why we _teach_ programming using Python.
(C) is the point about Python being easier to _learn_ than other mature alternatives.
I think (C) is important enough to justify spending a paragraph on it.

> How are the semantics of Python different than other languages?

In Python `"1" + 2` is valid syntactically but not semantically.
In Perl, the same expression is valid syntactically and semantically (result is 3).

In Python `"1" + "2"` is valid syntactically and semantically.
In Perl, the identical expression (`"1" + "2"`) is valid syntactically but not semantically (result is 0) and syntax of a semantically valid equivalent is `"1" . "2"`.

So, it's always a combination of syntax and semantics that makes a language intuitive or... not-so-much. I think that the closer the syntax and semantics of a programming language to those of (human language + mathematics), the more intuitive it is. 

> What makes for Python being "easy-to-learn syntax and semantics"? It may not be considered easy to someone, especially someone who has no reference to other programming syntax.

Agreed, this was not the best wording -- we don't have to use it and we can keep "It's easier to learn than most of the mature alternatives". I'm mostly concerned with overloading the last point as I tried to explain in the beginning of this comment.

ldko replied:
Perhaps you could do a pull request to submit your changes there @maxim-belkin if you are wanting to reword/reorganize beyond what is presented in this pull request . I think we may have lost @ndilekli :/.

ndilekli replied:
Hello, sorry that I was out of the loop for a while. I could do another commit, whenever you would like me to. Regarding the Python syntax, I could comment that as someone starting out with Basic, then continuing with Pascal and C, the syntax of Python is indeed much more like reading a text in English. Saying "reads like English" might be an overstatement, but I have seen quite a lot of remarks on that aspect. We might perhaps instead say it is more readable relative to many other programming languages like C and Java. What do you think?

ndilekli replied:
Hello, sorry that I was out of the loop for a while. I could do another commit, whenever you would like me to. Regarding the Python syntax, I could comment that as someone starting out with Basic, then continuing with Pascal and C, the syntax of Python is indeed much more like reading a text in English. Saying "reads like English" might be an overstatement, but I have seen quite a lot of remarks on that aspect. We might perhaps instead say it is more readable relative to many other programming languages like C and Java. What do you think?

maxim-belkin replied:
> Hello, sorry that I was out of the loop for a while. 

No worries!

> I could do another commit, whenever you would like me to.

Feel free to update the PR whenever you like if you agree with the comments / suggestions. To sum up:

1. My concern is that we are overloading the last item in the list that explains why we use Python for teaching with information why Python is easier to learn than other (mature) programming languages. Those are related but different topics. So, I propose to move the explanation to the paragraph right after the list. I'm not sure if Lauren (@ldko) agrees with that, so I'd like to clarify this first. If there is no consensus, we might have to leave it as is and discuss this part later in a separate PR or issue.

2. Please have a look at Lauren's suggestion on how to rephrase the first paragraph in introduction: https://github.com/swcarpentry/python-novice-inflammation/pull/837#discussion_r450963457. You can use it "as is" or make modifications you deem important.

>  as someone starting out with Basic, then continuing with Pascal and C

Huh, I took the same path with a quick stop at Visual Basic between Pascal and C (it was interesting to observe how Python adopted and extended Pascal's `:=` assignment operator).


> , the syntax of Python is indeed much more like reading a text in English. Saying "reads like English" might be an overstatement, but I have seen quite a lot of remarks on that aspect. We might perhaps instead say it is more readable relative to many other programming languages like C and Java.

Sure, that'll work too (I think). This way we won't have to go deep into syntax/semantics stuff.

Thank you, Naci!

ldko replied:
If you all agree about splitting off some of the last item in the list to a paragraph that follows, that is fine with me. Please go ahead.

ldko commented on _extras/guide.md:
```diff
@@ -27,13 +27,14 @@ is included as [`code/gen_inflammation.py`](../code/gen_inflammation.py).
 ## Overall
 
 This lesson is written as an introduction to Python,
```
I don't think a `.` makes sense there @maxim-belkin as the sentence continues after the punctuation.

maxim-belkin replied:
This proposed change is related to the one below: https://github.com/swcarpentry/python-novice-inflammation/pull/837#discussion_r450181162

maxim-belkin replied:
Resolving in favor of https://github.com/swcarpentry/python-novice-inflammation/pull/837#discussion_r450963457

ldko commented on _extras/guide.md:
```diff
@@ -27,13 +27,14 @@ is included as [`code/gen_inflammation.py`](../code/gen_inflammation.py).
-we must teach people a little about
-the mechanics of manipulating data with lists and file I/O
-so that their functions can do things they actually care about.
+but its wider purpose is to introduce an important idea in programming:
+how to solve problems by building functions, units of code that allow
+reusability and efficient logical organization of software.
```
I don't really think the sentence needs to change further, but if you are wanting to break it up and change it, I prefer "However" over "Still" and a `,` instead of "and" between "efficient" and "logical". However, I am fine with Maxim's suggestion as-if if that is what you all want. @ndilekli what please commit as you see fit!

maxim-belkin replied:
I'm OK if we don't break it into two but I think we need to add a comma or an `and` between "efficient" and "logical" and maybe change comma dashes before "the units". The reason I proposed to break this sentence up is that it seems to have too many "stops" / "breaks":
1. `, but`
2. `programming:`
3. `functions,`
4. `efficient [and/,] logical`

... and breaking up the sentence into two had minimal effect on the changes in this PR. Other way is to remove "to introduce an important idea in programming:" so that it reads

> but its wider purpose is to show how to solve problems by building functions...

but this is open up for a debate and, again, we can always leave it as-is here and address it in a future PR.

maxim-belkin replied:
Resolving in favor of https://github.com/swcarpentry/python-novice-inflammation/pull/837#discussion_r450963457
